[DependencyInjection] Preload on php 8

**Symfony version(s) affected**: 4.4.* and i think 5.* to.

**Description**  
```PHP
PHP Fatal error:  Uncaught Error: Call to undefined method ReflectionUnionType::is          
 Builtin() in /php-app/vendor/symfony/dependency-injection/Dumper/Preloader          
 .php:97                                                                                                                
 Stack trace:                                                                                                           
 #0 /php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(85):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::preloadType()                                                  
 #1 /php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(81):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::doPreload()                                                    
 #2 /php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(103):          
  Symfony\Component\DependencyInjection\Dumper\Preloader::doPreload()                                                   
 #3 /php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(85):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::preloadType()                                                  
 #4 /php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(41):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::doPreload()                                                    
 #5 /php-app/apps/web/var/cache/prod/srcApps_Web_KernelProdContainer.preloa          
 d.php(714): Symfony\Component\DependencyInjection\Dumper\Preloader::preload()                                          
 #6 /php-app/apps/web/src/.preload.php(4): require('...')                            
 #7 {main}                                                                                                              
   thrown in /php-app/vendor/symfony/dependency-injection/Dumper/Preloader.          
 php on line 97
```

**How to reproduce**  
PHP 8.0.0rc1 + symfony 4.4.15

With symfony cli:
`symfony new --version=lts --full sf4php8 && cd sf4php8 && echo '8.0.0' > .php-version && symfony serve`

**Possible Solution**  
<!--- Optional: only if you have suggestions on a fix/reason for the bug -->

**Additional context**  
In PHP 8, ReflectionType does not have an isBuiltin() method.

There is issue in on bugs.php.net related to this.
[https://bugs.php.net/bug.php?id=80247](https://bugs.php.net/bug.php?id=80247)

Actually i dont use them explicity in the code, only in the phpdocs (and this is allowed in any php version). I'am just trying to run same project which is sf 4.4 which works fine on php7.4 but fails on php8. Check bug on php.net because on php8 this behavior is different and there is no method isBuiltin() on ReflectionType, only on ReflectionNamedType class which is subtype of ReflectionType.

[https://github.com/symfony/symfony/blob/d6cc026bc1fa3a7b126357103c3c2fc74595608b/src/Symfony/Component/DependencyInjection/Dumper/Preloader.php#L97](https://github.com/symfony/symfony/blob/d6cc026bc1fa3a7b126357103c3c2fc74595608b/src/Symfony/Component/DependencyInjection/Dumper/Preloader.php#L97)

U can try it with empty symfony 4.4 project, it fails without any custom code:

`symfony new --version=lts --full sf4php8 && cd sf4php8 && echo '8.0.0' > .php-version && symfony serve`

```php
[ERROR] PHP failed to start:                                                                                           
 [Fri Oct 23 18:57:37 2020] PHP Fatal error:  Uncaught Error: Call to undefined method ReflectionUnionType::is          
 Builtin() in php-app/vendor/symfony/dependency-injection/Dumper/Preloader          
 .php:97                                                                                                                
 Stack trace:                                                                                                           
 #0 php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(85):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::preloadType()                                                  
 #1 php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(81):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::doPreload()                                                    
 #2 php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(103):          
  Symfony\Component\DependencyInjection\Dumper\Preloader::doPreload()                                                   
 #3 php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(85):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::preloadType()                                                  
 #4 php-app/vendor/symfony/dependency-injection/Dumper/Preloader.php(41):           
 Symfony\Component\DependencyInjection\Dumper\Preloader::doPreload()                                                    
 #5 php-app/apps/web/var/cache/prod/srcApps_Web_KernelProdContainer.preloa          
 d.php(714): Symfony\Component\DependencyInjection\Dumper\Preloader::preload()                                          
 #6 php-app/apps/web/src/.preload.php(4): require('...')                            
 #7 {main}                                                                                                              
   thrown in php-app/vendor/symfony/dependency-injection/Dumper/Preloader.          
 php on line 97          
```

This is beacuse of: 
https://bugs.php.net/bug.php?id=80247

It should be handled different for php7/php8.

--- a/.github/patch-types.php
+++ b/.github/patch-types.php
@@ -33,6 +33,7 @@ foreach ($loader->getClassMap() as $class => $file) {
         case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php'):
         case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/uniontype_classes.php'):
         case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/ParentNotExists.php'):
+        case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/'):
         case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Prototype/BadClasses/MissingParent.php'):
         case false !== strpos($file, '/src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/'):
         case false !== strpos($file, '/src/Symfony/Component/ErrorHandler/Tests/Fixtures/'):

--- a/src/Symfony/Component/DependencyInjection/Dumper/Preloader.php
+++ b/src/Symfony/Component/DependencyInjection/Dumper/Preloader.php
@@ -94,7 +94,7 @@ class Preloader
 
     private static function preloadType(?\ReflectionType $t, array &$preloaded): void
     {
-        if (!$t || $t->isBuiltin()) {
+        if (!$t) {
             return;
         }
 

--- a/src/Symfony/Component/DependencyInjection/Tests/Dumper/PreloaderTest.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Dumper/PreloaderTest.php
@@ -1,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Dumper;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\DependencyInjection\Dumper\Preloader;
+
+class PreloaderTest extends TestCase
+{
+    /**
+     * @requires PHP 7.4
+     */
+    public function testPreload()
+    {
+        $r = new \ReflectionMethod(Preloader::class, 'doPreload');
+        $r->setAccessible(true);
+
+        $preloaded = [];
+
+        $r->invokeArgs(null, ['Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\Dummy', &$preloaded]);
+
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\Dummy', false));
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\A', false));
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\B', false));
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\C', false));
+    }
+
+    /**
+     * @requires PHP 8
+     */
+    public function testPreloadUnion()
+    {
+        $r = new \ReflectionMethod(Preloader::class, 'doPreload');
+        $r->setAccessible(true);
+
+        $preloaded = [];
+
+        $r->invokeArgs(null, ['Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\UnionDummy', &$preloaded]);
+
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\UnionDummy', false));
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\D', false));
+        self::assertTrue(class_exists('Symfony\Component\DependencyInjection\Tests\Fixtures\Preload\E', false));
+    }
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/A.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/A.php
@@ -1,0 +1,16 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class A
+{
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/B.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/B.php
@@ -1,0 +1,16 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class B
+{
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/C.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/C.php
@@ -1,0 +1,16 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class C
+{
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/D.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/D.php
@@ -1,0 +1,16 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class D
+{
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/Dummy.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/Dummy.php
@@ -1,0 +1,30 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class Dummy
+{
+    public A $a;
+
+    public function doSomething(B $b): ?C
+    {
+        return null;
+    }
+
+    public function noTypes($foo)
+    {
+    }
+
+    public function builtinTypes(int $foo): ?string
+    {
+    }
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/E.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/E.php
@@ -1,0 +1,16 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class E
+{
+}

--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/UnionDummy.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/Preload/UnionDummy.php
@@ -1,0 +1,21 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Preload;
+
+final class UnionDummy
+{
+    public D|E $de;
+
+    public function builtinTypes(int|float $foo): string|\Stringable|null
+    {
+    }
+}

